src/bin/pg_autoctl/pgsql.c

```diff
@@ -165,6 +205,115 @@ pgsql_open_connection(PGSQL *pgsql)
 }
 
 
+/*
+ * pgsql_retry_open_connection loops over a PQping call until the remote server
+ * is ready to accept connections, and then connects to it and returns true
+ * when it could connect, false otherwise.
+ */
+static PGconn *
+pgsql_retry_open_connection(PGSQL *pgsql)
+{
+	PGconn *connection = NULL;
+	int attempts = 0;
+	bool retry = true;
+	bool connectionOk = false;
+	uint64_t startTime = time(NULL);
+
+	log_warn("Failed to connect to \"%s\", retrying until "
+			 "the server is ready", pgsql->connectionString);
+
+	while (retry)
+	{
+		uint64_t now = time(NULL);
+
+		if ((now - startTime) >= POSTGRES_PING_RETRY_TIMEOUT)
+		{
+			log_warn("Failed to connect to \"%s\" after %d attempts, "
+					 "stopping now", pgsql->connectionString, attempts);
+			break;
+		}
+
+		++attempts;
+		switch (PQping(pgsql->connectionString))
+		{
+			case PQPING_OK:
+			{
+				log_debug("PQping OK after %d attempts", attempts);
+				retry = false;
+
+				/*
+				 * Ping is now ok, and connection is still NULL because the
```

Comment:
I'm not familiar with pgping, but I suppose it's possible that after connecting the database is still not ready yet. Sometimes Postgres will let you connect but respond to queries with an error like "the database engine is starting up" or "the database engine is shutting down" or (if the query makes a change) "the database is in a read-only mode" (if it is performing recovery as part of startup).

In our own code we wrap not only connecting but also running a query in the retry loop and consider those transient errors from Postgres to be reasons to retry.